chain/chunks/src/lib.rs

```diff
@@ -178,53 +204,124 @@ impl SealsManager {
+
+                let chosen = candidates
+                    .choose_multiple(
+                        &mut rand::thread_rng(),
+                        cmp::min(NUM_PARTS_REQUESTED_IN_SEAL, candidates.len()),
+                    )
+                    .cloned()
+                    .collect::<HashSet<_>>();
+                let challenge = ActiveSealChallenge {
+                    part_ords: chosen,
+                    chunk_producer,
+                    sent: Utc::now(),
+                    height,
+                };
+
+                Ok(entry.insert(challenge))
             }
-            let chosen = candidates
-                .choose_multiple(
-                    &mut rand::thread_rng(),
-                    cmp::min(NUM_PARTS_REQUESTED_IN_SEAL, candidates.len()),
-                )
-                .cloned()
-                .collect::<HashSet<_>>();
-            Seal { part_ords: chosen, chunk_producer, sent: Utc::now() }
-        }))
+        }
     }
 
     fn approve_chunk(&mut self, chunk_hash: &ChunkHash) {
-        let seal = self.seals.get_mut(chunk_hash).expect("seal should be already produced");
-        seal.part_ords.clear();
+        let seal =
+            self.active_challenges.remove(chunk_hash).expect("seal should be already produced");
+        Self::insert_previous_seal(&mut self.previous_seals, seal.height, chunk_hash.clone());
     }
 
-    fn track_seals(&mut self) {
-        let now = Utc::now();
-        for (chunk_hash, seal) in self.seals.iter_mut() {
-            if seal.part_ords.len() > NUM_PARTS_LEFT_IN_SEAL
-                && (now - seal.sent).num_milliseconds() > ACCEPTING_SEAL_PERIOD_MS
-            {
-                warn!(target: "client", "Couldn't reconstruct chunk {:?} from {:?}, I'm {:?}", chunk_hash, seal.chunk_producer, self.me);
-                self.dont_include_chunks_from.insert(seal.chunk_producer.clone());
-                seal.part_ords.clear();
+    fn insert_previous_seal(
+        previous_seals: &mut BTreeMap<BlockHeight, HashSet<ChunkHash>>,
+        height: BlockHeight,
+        chunk_hash: ChunkHash,
+    ) {
+        let hashes_at_height = previous_seals.entry(height).or_insert_with(HashSet::new);
+        hashes_at_height.insert(chunk_hash);
+    }
+
+    fn prune_previous_seals(&mut self, block_hash: &CryptoHash) {
+        if let Ok(gc_stop_height) = self.runtime_adapter.get_gc_stop_height(block_hash) {
+            if let Some(least_height) = self.previous_seals.keys().next() {
+                if least_height < &gc_stop_height {
+                    let remaining_seals = self.previous_seals.split_off(&gc_stop_height);
+                    self.previous_seals = remaining_seals;
+                }
             }
         }
     }
 
-    fn should_trust_chunk_producer(&self, chunk_producer: &AccountId) -> bool {
-        !self.dont_include_chunks_from.contains(chunk_producer)
+    fn track_seals(&mut self, block_hash: &CryptoHash) {
+        let now = Utc::now();
```

Comment:
Are we going to track seals not often than per 1 second?

Reply:
I don't think that optimization is needed and that it will therefore only introduce unnecessary complexity. The reason I think it is not needed is because in master presently the `track_seals` loop is over all seals that have been created ever, and to my knowledge it is not a performance bottleneck. If the nodes can get by with a week's worth of seals (that is how often we hard fork presently right?) then I am sure they will be ok if we only let each element live in the active collection for 30 seconds.